rust-src/bulletproofs/src/utils.rs

```diff
@@ -0,0 +1,132 @@
+use crypto_common::*;
+use crypto_common_derive::*;
+use curve_arithmetic::Curve;
+use ff::Field;
+#[cfg(test)]
+use rand::Rng;
+/// Struct containing generators G and H needed for range proofs
+#[allow(non_snake_case)]
+#[derive(Debug, Clone, Serialize, SerdeBase16Serialize)]
+pub struct Generators<C: Curve> {
+    #[size_length = 4]
+    pub G_H: Vec<(C, C)>,
+}
+
+impl<C: Curve> Generators<C> {
+    /// **Warning** do not use in production!
+    /// This **unsafely** generates a list of generators of a given size for
+    /// testing purposes. For production, generator must be created with
+    /// care.
+    #[cfg(test)]
+    pub fn generate(n: usize, csprng: &mut impl Rng) -> Self {
+        let mut gh = Vec::with_capacity(n);
+        for _ in 0..n {
+            let x = C::generate(csprng);
+            let y = C::generate(csprng);
+            gh.push((x, y));
+        }
+        Self { G_H: gh }
+    }
+
+    pub fn take(&self, nm: usize) -> Self {
+        Self {
+            G_H: self.G_H[0..nm].to_vec(),
+        }
+    }
+}
+
+/// This function takes one argument n and returns the
+/// vector (z^j, z^{j+1}, ..., z^{j+n-1}) in F^n for any field F
+/// The arguments are
+/// - z - the field element z
+/// - first_power - the first power j
+/// - n - the integer n.
+pub fn z_vec<F: Field>(z: F, first_power: usize, n: usize) -> Vec<F> {
+    let mut z_n = Vec::with_capacity(n);
+    // let mut z_i = F::one();
+    // FIXME: This should would be better to do with `pow`.
+    // for _ in 0..first_power {
+    //    z_i.mul_assign(&z);
+    //}
+    let exp: [u64; 1] = [first_power as u64];
+    let mut z_i = z.pow(exp);
+    for _ in 0..n {
+        z_n.push(z_i);
+        z_i.mul_assign(&z);
+    }
+    z_n
+}
+
+/// Converts the u64 set vector into a vector over the field
+pub fn get_set_vector<C: Curve>(the_set: &[u64]) -> Vec<C::Scalar> {
+    let n = the_set.len();
+    let mut s_vec = Vec::with_capacity(n);
+    for elem_i in the_set {
+        let s_i = C::scalar_from_u64(*elem_i);
+        s_vec.push(s_i);
+    }
+    s_vec
+}
+
+/// Pads a field vector two a power of two length by repeating the last element
```

Comment:
That's not entirely what it does though, right. If the vector is empty nothing is done. Document this behaviour.
```suggestion
/// Pads a field vector to a power of two length by repeating the last element
```